File: tests/matrix/delocalize/matrix_delocalize.cpp

#include "../../../include/optitrust.h"

typedef int T;

T* b;

const int N0 = 5;
const int N1 = 10;
const int N2 = 10;
const int N3 = 10;

void allocate () {

  b = (T*) CALLOC3 (N1, N2, N3, sizeof(T));
}

int main(){

  T* a = (T*) CALLOC3 (N1, N2, N3, sizeof(T));
  for (int i = 0; i < 10; i++) {
    int t = a[MINDEX3(N1,N2,N3,i,i+1,i+2)];
    a[MINDEX3(N1,N2,N3,i,i+1,i+2)] = t + 5;
  }

  for (int j = 0; j < 10; j++) {
    b[MINDEX3(N1,N2,N3,j,j+1,j+2)];
  }

  int z = 0;
  return 0;
}

File: tests/matrix/delocalize/matrix_delocalize.ml
open Optitrust
open Target

let _ = Run.script_cpp (fun _ ->

   !! Matrix.delocalize "a" ~into:"x" ~init_zero:true ~acc_in_place:false  ~acc:"sum" ~dim:(var "N0") ~index:"i0" ~ops:(Local_arith (Lit_int 0, Binop_add)) [cFor "i"];

   let alloc_instr = [cTopFunDef "allocate"; cWriteVar "b"] in
   !! Matrix.delocalize "b" ~into:"y" ~init_zero:true ~acc_in_place:false  ~acc:"sum" ~dim:(var "N0") ~index:"i0" ~ops:(Local_arith (Lit_int 0, Binop_add)) ~alloc_instr ~labels:["alloc";"";"dealloc"] [cFor "j"];

   !! Trace.alternative (fun () ->
        !! Matrix.delocalize "a" ~into:"x" ~last:true  ~init_zero:true ~acc_in_place:false  ~acc:"sum" ~dim:(var "N0") ~index:"i0" ~ops:(Local_arith (Lit_int 0, Binop_add)) [cFor "i"];
        !!());

  !! Trace.alternative (fun () ->
        !! Matrix.delocalize "a" ~use:(expr "k") ~into:"x" ~init_zero:true ~acc_in_place:false  ~acc:"sum" ~dim:(var "N0") ~index:"i0" ~ops:(Local_arith (Lit_int 0, Binop_add)) [cFor "i"];
        !!());

)
